nixos/acme: improve scalability - reduce superfluous unit activations

ctheune opened the pull request:
##  acme: improve scalability - reduce superfluous unit activations

The previous setup caused all renewal units to be triggered upon
ever so slight changes in config. In larger setups (100+ certificates)
adding a new certificate caused high system load and/or large memory
consumption issues. The memory issues are already a alleviated with
the locking mechanism. However, this then causes long delays upwards
of multiple minutes depending on individual runs and also caused
superfluous activations.

In this change we streamline the overall setup of units:

1. The unit that other services can depend upon is 'acme-{cert}.service'.
We call this the 'base unit'. As this one as `RemainAfterExit` set
the `acme-finished-{cert}` targets are not required any longer.

2. We now always generate initial self-signed certificates to simplify
the dependency structure. This deprecates the `preliminarySelfsigned`
option.

3. The `acme-order-renew-{cert}` service gets activated after the base
unit and services using certificates have started and performs all acme
interactions. When it finishes others services (like web servers) will
be notified through the `reloadServices` option or they can use
`wantedBy` and `after` dependencies if they implement their own reload
units.

The renewal timer also triggers this unit.

4. The timer unit is explicitly blocked from being started by s-t-c.

5. Permission management has been cleaned up a bit: there was an
   inconsistency between having the .lego files set to 600 vs 640
   on the exposed side. This is unified to 640 now.

6. Exempt the account target from being restarted by s-t-c. This will
   happen automatically if something relevant to the account changes.

##  acme: switch concurrency limit to a runtime-based implementation

The previous implementation caused triggers on many units when adding
or removing certificates because the baked-in lock file assignments
changed.

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

arianvp commented on nixos/modules/security/acme/default.nix:
```diff
@@ -1177,15 +1180,26 @@ in
               lib.mapAttrs (_: f: f null) selfsignServiceFunctions;
+          # This indirection is needed because the actual renewal units remain
+          # active so that s-t-c can update them based on changes.
+          triggerRenewServices = lib.mapAttrs' (
```
It might make sense to split up  into two services.  One for `lego run` (request new certificate) and one for `lego renew`.

`lego renew` only runs through timer. `lego run` can instead by pulled in by `multi-user.target` (or whatever service pulls it in).

Then I think the stuff here gets a bit clearer

m1cr0man reviewed (changes requested):
Thanks for doing all this work. It looks great :) I have some general comments but nothing too major.

ctheune commented on nixos/tests/acme/webserver.nix:
```diff
@@ -2,7 +2,7 @@
   group,
   baseModule,
-  domain ? "example.test",
+  domain,
```
Because all the specific tests *need* to know the domain as their integration points into the test refer to them but there's no way to access that variable for integration points so I opted to have it set at the higher level in `nixos/tests/acme/default.nix`.

ctheune commented on nixos/modules/services/web-servers/caddy/default.nix:
```diff
@@ -411,11 +409,8 @@ in
-        ++ map (certName: "acme-${certName}.service") independentCertNames; # avoid loading self-signed key w/ real cert, or vice-versa
-      before = map (certName: "acme-${certName}.service") dependentCertNames;
+      wants = map (certName: "acme-${certName}.service") vhostCertNames;
+      after = map (certName: "acme-${certName}.service") vhostCertNames;
```
I've kept httpd and nginx's config in sync, but caddy was structured in a completely different fashion. As the tests pass I think it's in a status-quo state comparable as to how it was before - and looking at the tests like this:

```
# FIXME reloading caddy is not sufficient to load new certs.
# Restart it manually until this is fixed.
```

it appears that the whole reload story with caddy wasn't actually properly done so far. Not sure whether this should be extended within this PR as it's already quite big. I think I haven't made it worse from before, though ...

ctheune commented on nixos/modules/services/web-servers/nginx/default.nix:
```diff
@@ -1578,29 +1575,28 @@ in
       source = configFile;
     };
 
+    # XXX replace by using the acme reload mechanism?
```
agree!

ctheune commented on nixos/modules/security/acme/default.nix:
```diff
@@ -210,7 +185,8 @@ let
         BindPaths = "/var/lib/acme/.minica:/tmp/ca";
       };
-    })
+    }
```
Agree

ctheune commented on nixos/modules/security/acme/default.nix:
```diff
@@ -553,24 +562,27 @@ let
-            exit 10
+          else
+            # Do a full run
+            if ! lego ${runOpts}; then
```
Yeah, that just got lost due to lots of back and forth while writing this. Agree.

ctheune committed:
nixos/acme: update release notes and documentation

ctheune committed:
nixos/acme: improve scalability - reduce superfluous unit activations

The previous setup caused all renewal units to be triggered upon
ever so slight changes in config. In larger setups (100+ certificates)
adding a new certificate caused high system load and/or large memory
consumption issues. The memory issues are already a alleviated with
the locking mechanism. However, this then causes long delays upwards
of multiple minutes depending on individual runs and also caused
superfluous activations.

In this change we streamline the overall setup of units:

1. The unit that other services can depend upon is 'acme-{cert}.service'.
We call this the 'base unit'. As this one as `RemainAfterExit` set
the `acme-finished-{cert}` targets are not required any longer.

2. We now always generate initial self-signed certificates to simplify
the dependency structure. This deprecates the `preliminarySelfsigned`
option.

3. The `acme-order-renew-{cert}` service gets activated after the base
unit and services using certificates have started and performs all acme
interactions. When it finishes others services (like web servers) will
be notified through the `reloadServices` option or they can use
`wantedBy` and `after` dependencies if they implement their own reload
units.

The renewal timer also triggers this unit.

4. The timer unit is explicitly blocked from being started by s-t-c.

5. Permission management has been cleaned up a bit: there was an
   inconsistency between having the .lego files set to 600 vs 640
   on the exposed side. This is unified to 640 now.

6. Exempt the account target from being restarted by s-t-c. This will
   happen automatically if something relevant to the account changes.

ctheune committed:
nixos/acme: switch concurrency limit to a runtime-based implementation

The previous implementation caused triggers on many units when adding
or removing certificates because the baked-in lock file assignments
changed.

m1cr0man commented on nixos/doc/manual/release-notes/rl-2511.section.md:
```diff
@@ -96,6 +96,9 @@
 - `services.clamsmtp` is unmaintained and was removed from Nixpkgs.
 
+- The option `security.acme.preliminarySelfsigned` has been removed to support a more streamlined
+  ACME unit dependency configuration.
```
Yep perfect :)

m1cr0man commented on nixos/modules/security/acme/default.nix:
```diff
@@ -365,67 +365,103 @@ let
           BindPaths = [
             "/var/lib/acme/.minica:/tmp/ca"
-            "/var/lib/acme/${cert}:/tmp/${keyName}"
+            "/var/lib/acme/${cert}:/tmp/out"
```
That seems reasonable.

m1cr0man reviewed (approved):
This is looking great. Thanks once again for all this work. :smile:

SuperSandro2000 commented on nixos/doc/manual/release-notes/rl-2511.section.md:
```diff
@@ -165,6 +165,21 @@
+
+  We now always generate initial self-signed certificates as this drastically simplifies the dependency structure. As a result, the option `security.acme.preliminarySelfsigned` has been removed.
+
+  Instead of the previous `acme-finished-{certname}.target`s there are now `acme-order-renew-{certname}` services that will be activated
```
```suggestion
  Instead of the previous `acme-finished-{certname}.target`s there are now `acme-order-renew-{certname}.service`'s that will be activated
```
That will probably make the change clearer.

leona-ya reviewed (approved):
Aside from Sandro's comment this seems good to me

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/security-acme-defaults-to-minica-instead-of-lego-with-cloudflare-dns-on-nixos-25-11/73220/4
